File: routes/jobs_routes.py
from flask import Blueprint, request, jsonify
from tasks import process_leetcode_task, process_test_case_task, process_code_modification_task
from celery.result import AsyncResult

jobs_bp = Blueprint('jobs', __name__)

@jobs_bp.route('/api/jobs/leetcode', methods=['POST'])
def submit_leetcode_job():
    """Submit a LeetCode problem for async processing"""
    from cache import cache

    problem_number = request.json.get('problem_number', '')
    custom_prompt = request.json.get('custom_prompt', None)

    # Submit task to Celery with current settings from Redis
    task = process_leetcode_task.delay(
        problem_number,
        custom_prompt,
        cache.get_current_model(),
        use_cache=cache.is_enabled(),
        model_aware_cache=cache.is_model_aware_cache()
    )

    return jsonify({
        'job_id': task.id,
        'status': 'submitted',
        'message': 'Job submitted for processing'
    }), 202

@jobs_bp.route('/api/jobs/test-cases', methods=['POST'])
def submit_test_case_job():
    """Submit test case generation for async processing"""
    from cache import cache

    code = request.json.get('code', '')

    # Submit task to Celery with current cache settings
    task = process_test_case_task.delay(
        code,
        cache.get_current_model(),
        use_cache=cache.is_enabled(),
        model_aware_cache=cache.is_model_aware_cache()
    )

    return jsonify({
        'job_id': task.id,
        'status': 'submitted',
        'message': 'Test case generation job submitted'
    }), 202

@jobs_bp.route('/api/jobs/code-modification', methods=['POST'])
def submit_code_modification_job():
    """Submit code modification for async processing"""
    from cache import cache

    prompt = request.json.get('prompt', '')
    code = request.json.get('code', '')

    # Submit task to Celery with current cache settings
    task = process_code_modification_task.delay(
        prompt,
        code,
        cache.get_current_model(),
        use_cache=cache.is_enabled(),
        model_aware_cache=cache.is_model_aware_cache()
    )

    return jsonify({
        'job_id': task.id,
        'status': 'submitted',
        'message': 'Code modification job submitted'
    }), 202

@jobs_bp.route('/api/jobs/<job_id>', methods=['GET'])
def get_job_status(job_id):
    """Check status of a background job"""
    task = AsyncResult(job_id)

    if task.state == 'PENDING':
        response = {
            'job_id': job_id,
            'state': task.state,
            'status': 'Job is waiting to be processed'
        }
    elif task.state == 'STARTED':
        response = {
            'job_id': job_id,
            'state': task.state,
            'status': 'Job is being processed'
        }
    elif task.state == 'SUCCESS':
        response = {
            'job_id': job_id,
            'state': task.state,
            'status': 'Job completed successfully',
            'result': task.result
        }
    elif task.state == 'FAILURE':
        response = {
            'job_id': job_id,
            'state': task.state,
            'status': 'Job failed',
            'error': str(task.info)
        }
    else:
        response = {
            'job_id': job_id,
            'state': task.state,
            'status': f'Job state: {task.state}'
        }

    return jsonify(response)

File: backend-node/routes/jobsRoutes.js
import express from 'express';
import { testCaseQueue, codeModificationQueue, leetcodeQueue, getJobStatus, addJobWithModel } from '../queue.js';

const router = express.Router();

// POST /api/jobs/leetcode
router.post('/jobs/leetcode', async (req, res) => {
  try {
    const { problem_number, custom_prompt } = req.body;

    const job = await addJobWithModel(leetcodeQueue, {
      problemNumber: problem_number,
      customPrompt: custom_prompt
    });

    res.status(202).json({
      job_id: job.id,
      status: 'submitted',
      message: 'Job submitted for processing'
    });
  } catch (error) {
    res.json({ success: false, error: error.message });
  }
});

// POST /api/jobs/test-cases
router.post('/jobs/test-cases', async (req, res) => {
  try {
    const { code } = req.body;

    const job = await addJobWithModel(testCaseQueue, {
      code
    });

    res.status(202).json({
      job_id: job.id,
      status: 'submitted',
      message: 'Test case generation job submitted'
    });
  } catch (error) {
    res.json({ success: false, error: error.message });
  }
});

// POST /api/jobs/code-modification
router.post('/jobs/code-modification', async (req, res) => {
  try {
    const { prompt, code } = req.body;

    const job = await addJobWithModel(codeModificationQueue, {
      prompt,
      code
    });

    res.status(202).json({
      job_id: job.id,
      status: 'submitted',
      message: 'Code modification job submitted'
    });
  } catch (error) {
    res.json({ success: false, error: error.message });
  }
});

// GET /api/jobs/:job_id
router.get('/jobs/:job_id', async (req, res) => {
  try {
    const { job_id } = req.params;

    // Try to find the job in all queues
    let jobStatus = null;

    // Try test-cases queue
    jobStatus = await getJobStatus(job_id, 'test-cases');
    if (jobStatus && jobStatus.state !== 'PENDING') {
      return res.json({
        job_id,
        state: jobStatus.state,
        status: jobStatus.status,
        result: jobStatus.result,
        error: jobStatus.error
      });
    }

    // Try code-modification queue
    jobStatus = await getJobStatus(job_id, 'code-modification');
    if (jobStatus && jobStatus.state !== 'PENDING') {
      return res.json({
        job_id,
        state: jobStatus.state,
        status: jobStatus.status,
        result: jobStatus.result,
        error: jobStatus.error
      });
    }

    // Try leetcode queue
    jobStatus = await getJobStatus(job_id, 'leetcode');
    if (jobStatus && jobStatus.state !== 'PENDING') {
      return res.json({
        job_id,
        state: jobStatus.state,
        status: jobStatus.status,
        result: jobStatus.result,
        error: jobStatus.error
      });
    }

    // Job not found in any queue
    res.json({
      job_id,
      state: 'PENDING',
      status: 'Job is waiting to be processed'
    });

  } catch (error) {
    res.json({ success: false, error: error.message });
  }
});

export default router;
